### PR TITLE
feat(qdrant): Add Qdrant Vector DB

### PR DESCRIPTION
**Pull request by EmanuelCampos**

### CODE REVIEW

File: packages/core/src/storage/vectorStore/QdrantVectorStore.ts [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+   */
+  async buildPoints(nodes: BaseNode[]): Promise<{
+    points: PointStruct[];
+    ids: string[];
+  }> {
+    const points: PointStruct[] = [];
+    const ids = [];
+
+    for (let i = 0; i < nodes.length; i++) {
+      const node_ids = [];
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/QdrantVectorStore.ts:139]: nitpik: no snakecase in TS

**Review by marcusschiesser — COMMENTED** [review]
@EmanuelCampos looking good. Is the data stored also interchangeable with the Python version. E.g. if you ingest data with the Python version, can you use it in LITS?

**Comment by EmanuelCampos** [comment]
@marcusschiesser Just dense vector should work. but it's without sparse vectors yet and hybrid search, but I think this additions can become later

**Review by marcusschiesser — APPROVED** [review]

**Comment by marcusschiesser** [comment]
@EmanuelCampos great, then this is ready to merge, please just add a changeset before (`pnpm changeset`)

File: packages/core/src/storage/vectorStore/types.ts [hunk 2]
```diff
@@ -49,6 +49,7 @@ export interface VectorStoreQuery {
   queryEmbedding?: number[];
   similarityTopK: number;
   docIds?: string[];
+  nodeIds?: string[];
```
Comment (yisding) [inline packages/core/src/storage/vectorStore/types.ts:52]: Here, docIds and nodeIds are duplicative.

Comment (yisding) [reply packages/core/src/storage/vectorStore/types.ts:52]: If you want, you can switch it globally to nodeIds and remove docIds..

Comment (EmanuelCampos) [reply packages/core/src/storage/vectorStore/types.ts:52]: @yisding Aren't

https://github.com/run-llama/LlamaIndexTS/blob/main/packages/core/src/Node.ts#L51

The nodeId is a random uuid

The document id is another one, that will be possible to update when document management be added

As well when you break the document in nodes they'll have the same document id but different nodeIds

Comment (marcusschiesser) [reply packages/core/src/storage/vectorStore/types.ts:52]: how about moving this out to another PR then?

Comment (EmanuelCampos) [reply packages/core/src/storage/vectorStore/types.ts]: @marcusschiesser looks good, removed it for now

**Review by marcusschiesser — APPROVED** [review]

**Comment by marcusschiesser** [comment]
@hiepxanh the PR is merged - looking forward to hearing your test report :D

**Comment by hiepxanh** [comment]
image

I follow the docs. the docs is working as expect. But the docs include port is not necessary. I'm doing more test on my project. It look very good

**Comment by marcusschiesser** [comment]
@hiepxanh thanks for the test and for finding this issue! Want to send a PR too?

**Comment by hiepxanh** [comment]
> @hiepxanh thanks for the test and for finding this issue! Want to send a PR too?

Yes it easy, I'm doing more test to make sure it work. But it doesn't work on large document. 

 I'm dealing with qdrant missmatch size, did I missing something in llamaindex or qdrant config? I read all the information but have no Idea? the llamaindex wrap everything in engine to index, I only see the size setting in create collection but not `fromDocument` seem trying to upsert node. do I need to change chunk size or something?

```
    const serviceContext = serviceContextFromDefaults({
        llm: new OpenAI({
            additionalSessionOptions: {
                baseURL: modelUrl,
            }
        }),
        embedModel: new OpenAIEmbedding({
            additionalSessionOptions: {
                baseURL: embeddingUrl
            }
        }),
        // chunkSize: 80,
        // chunkOverlap: 0,
    });

    const index = await VectorStoreIndex.fromDocuments(
        [document],
        {
            serviceContext,
            logProgress: true,
            vectorStore,
        });


    const queryEngine = index.asQueryEngine();

```

the client log:

```
Finished parsing documents.
Getting embedding for node 1/35
Getting embedding for node 2/35
Getting embedding for node 3/35
Getting embedding for node 4/35
Getting embedding for node 5/35
Getting embedding for node 6/35
Getting embedding for node 7/35
Getting embedding for node 8/35
Getting embedding for node 9/35
Getting embedding for node 10/35
Getting embedding for node 11/35
Getting embedding for node 12/35
Getting embedding for node 13/35
Getting embedding for node 14/35
Getting embedding for node 15/35
Getting embedding for node 16/35
Getting embedding for node 17/35
Getting embedding for node 18/35
Getting embedding for node 19/35
Getting embedding for node 20/35
Getting embedding for node 21/35
Getting embedding for node 22/35
Getting embedding for node 23/35
Getting embedding for node 24/35
Getting embedding for node 25/35
Getting embedding for node 26/35
Getting embedding for node 27/35
Getting embedding for node 28/35
Getting embedding for node 29/35
Getting embedding for node 30/35
Getting embedding for node 31/35
Getting embedding for node 32/35
Getting embedding for node 33/35
Getting embedding for node 34/35
Getting embedding for node 35/35


C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+openapi-typescript-fetch@1.2.1\node_modules\@qdrant\openapi-typescript-fetch\dist\cjs\fetcher.js:150
                throw new fun.Error(err);
                      ^
Error: Bad Request
    at Object.fun [as upsertPoints] (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+openapi-typescript-fetch@1.2.1\node_modules\@qdrant\openapi-typescript-fetch\dist\cjs\fetcher.js:150:23)
```


the database log


```

2024-01-26T05:06:20.293981Z  WARN collection::collection_manager::collection_updater: Update operation declined: Wrong input: Vector inserting error: expected dim: 1536, got 384    
2024-01-26T05:06:20.294140Z  INFO actix_web::middleware::logger: 172.19.0.1 "PUT /collections/default/points?wait=true HTTP/1.1" 400 110 "-" "qdrant-js" 0.168578    

```

**Comment by hiepxanh** [comment]
ok I found this line:
https://github.com/run-llama/LlamaIndexTS/blob/3a29a8036b15ad6e768a38f451dd17120e87fa1b/packages/core/src/storage/vectorStore/QdrantVectorStore.ts#L180C9-L180C50

```
 async add(embeddingResults: BaseNode[]): Promise<string[]> {
    if (embeddingResults.length > 0 && !this._collectionInitialized) {
      await this.initializeCollection(
        embeddingResults[0].getEmbedding().length, <-- this one should work but have no idea why it not 
      );
    }

    const { points, ids } = await this.buildPoints(embeddingResults);
```

**Comment by hiepxanh** [comment]
@marcusschiesser  ok I get over this issue:


https://github.com/run-llama/LlamaIndexTS/blob/3a29a8036b15ad6e768a38f451dd17120e87fa1b/packages/core/src/storage/vectorStore/QdrantVectorStore.ts#L72C5-L72C55

collection name default is "default".

|    `this.collectionName = collectionName ?? "default";`

1) if I test first time with test string => collection empty => create default using 1536 dimension
2) I use embedding model => collection exist, llamaindex reuse => reuse collection default (1536) with data 384
3) get error


i fix it by add new collection name:

```
const vectorStore = new QdrantVectorStore({
    url: "http://localhost:6333",
    collectionName: 'qdrant-test' <--- new name to create new collection to fix error
});

```

It seem work now. What is wrong with the behavior? Should we create warning or something? I think some kind of check and compare dimention or catch error and notice about this?

or write document about this to warning people

**Comment by marcusschiesser** [comment]
@hiepxanh great catch. My educated guess is that you created the collection first using the OpenAI embedding (dimension 1536) and then you're using another embedding model with dimension 384. 

@EmanuelCampos I guess we can throw an error in the `add` function if the dimension of the embedding doesn't fit the one in the collection?

**Comment by hiepxanh** [comment]
@marcusschiesser I added new push include the fix I mention above and known issues section. 

@EmanuelCampos  I guess the upsert is here, what do you think about this? I will do update on the PR include?

https://github.com/run-llama/LlamaIndexTS/blob/fe03aaae556e98263f58e37d1b1fd30307e99aae/packages/core/src/storage/vectorStore/QdrantVectorStore.ts#L186
```
const batchUpsert = async (points: PointStruct[]) => {
      await this.db.upsert(this.collectionName, {
        points: points,
      });
    };
```

```typescript
const batchUpsert = async (points: PointStruct[]) => {
      try {
        await this.db.upsert(this.collectionName, {
          points: points,
        });
      } catch (error: any) {
        console.error(error);
        throw new Error("QdrantVectorStore upsert error: " + error.message);
      }
    };
```

from this:

```
Getting embedding for node 33/35
Getting embedding for node 34/35
Getting embedding for node 35/35


C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+openapi-typescript-fetch@1.2.1\node_modules\@qdrant\openapi-typescript-fetch\dist\cjs\fetcher.js:150
                throw new fun.Error(err);
                      ^
Error: Bad Request
    at Object.fun [as upsertPoints] (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+openapi-typescript-fetch@1.2.1\node_modules\@qdrant\openapi-typescript-fetch\dist\cjs\fetcher.js:150:23)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at QdrantClient.upsert (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+js-client-rest@1.7.0_typescript@5.3.3\node_modules\@qdrant\js-client-rest\dist\cjs\qdrant-client.js:534:26)


>  NX  Process exited with code 1, waiting for changes to restart...
```
to this

```
Getting embedding for node 34/35
Getting embedding for node 35/35
ApiError: Bad Request
    at Object.fun [as upsertPoints] (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+openapi-typescript-fetch@1.2.1\node_modules\@qdrant\openapi-typescript-fetch\dist\cjs\fetcher.js:150:23)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at QdrantClient.upsert (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\@qdrant+js-client-rest@1.7.0_typescript@5.3.3\node_modules\@qdrant\js-client-rest\dist\cjs\qdrant-client.js:534:26) {
  headers: _HeadersList {
    cookies: null,
    [Symbol(headers map)]: Map(5) {
      'transfer-encoding' => [Object],
      'content-type' => [Object],
      'content-encoding' => [Object],
      'vary' => [Object],
      'date' => [Object]
    },
    [Symbol(headers map sorted)]: null
  },
  url: 'http://localhost:6333/collections/default/points?wait=true',
  status: 400,
  statusText: 'Bad Request',
  data: {
    status: {
      error: 'Wrong input: Vector inserting error: expected dim: 1536, got 384'
    },
    time: 0.14764534
  }
}



C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\llamaindex@0.1.0_typescript@5.3.3\node_modules\llamaindex\dist\index.js:5282
        var info = gen[key](arg);
                           ^
Error: QdrantVectorStore error: Bad Request
    at C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\llamaindex@0.1.0_typescript@5.3.3\node_modules\llamaindex\dist\index.js:5429:27
    at Generator.throw (<anonymous>)
    at asyncGeneratorStep$F (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\llamaindex@0.1.0_typescript@5.3.3\node_modules\llamaindex\dist\index.js:5282:28)
    at _throw (C:\Users\hiepx\small-cosmos\losa\node_modules\.pnpm\llamaindex@0.1.0_typescript@5.3.3\node_modules\llamaindex\dist\index.js:5303:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)


>  NX  Process exited with code 1, waiting for changes to restart...
```

**Pull request merged**